vertx-jdbc-client Oracle 12c - operation not allowed

The vert.x jdbc client is throwing a "SQLException: operation not allowed" when performing an update on an Oracle 12c database using the vertx-jdbc-client 3.2.1. When performing an updateWithParams or update and passing in a simple update statement it fails with the following error:

java.lang.RuntimeException: java.sql.SQLException: operation not allowed
at com.sample.verticle.MyVerticle.lambda$8(MyVerticle.java:192)
at com.sample.verticle.MyVerticle$$Lambda$43/984091703.handle(Unknown Source)
at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction.lambda$null$8(AbstractJDBCAction.java:65)
at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction$$Lambda$41/2139817074.handle(Unknown Source)
at io.vertx.core.impl.ContextImpl.lambda$wrapTask$18(ContextImpl.java:335)
at io.vertx.core.impl.ContextImpl$$Lambda$23/378909131.run(Unknown Source)
at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:358)
at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357)
at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:112)
at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.SQLException: operation not allowed
at oracle.jdbc.driver.OracleStatement.getGeneratedKeys(OracleStatement.java:5487)
at oracle.jdbc.driver.OracleStatementWrapper.getGeneratedKeys(OracleStatementWrapper.java:380)
at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.getGeneratedKeys(NewProxyPreparedStatement.java:2570)
at io.vertx.ext.jdbc.impl.actions.JDBCUpdate.execute(JDBCUpdate.java:55)
at io.vertx.ext.jdbc.impl.actions.JDBCUpdate.execute(JDBCUpdate.java:35)
at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction.handle(AbstractJDBCAction.java:52)
at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction.lambda$execute$10(AbstractJDBCAction.java:68)
at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction$$Lambda$39/1411764370.handle(Unknown Source)
at io.vertx.core.impl.ContextImpl.lambda$wrapTask$18(ContextImpl.java:335)
at io.vertx.core.impl.ContextImpl$$Lambda$23/378909131.run(Unknown Source)
at io.vertx.core.impl.OrderedExecutorFactory$OrderedExecutor.lambda$new$265(OrderedExecutorFactory.java:91)
at io.vertx.core.impl.OrderedExecutorFactory$OrderedExecutor$$Lambda$20/988705146.run(Unknown Source)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
... 1 mo

Here's what I'm doing:
```java
String sql = "update Whisky w set w.name='AAAAAAA' where w.id = ?"; 
    JsonArray params = new JsonArray().add(1); 
    this.client.getConnection(conn -> { 
    if (conn.succeeded()) { 
        SQLConnection connection = conn.result(); 
        connection.updateWithParams(sql, params, query -> { 
            if (query.succeeded()) { 
                System.out.println("Updated row"); 
            } 
            else { 
                throw new RuntimeException(query.cause()); 
        } 
        connection.close();
 }); 
} 
}); 
````
I posted this as a question here:
https://groups.google.com/forum/#!topic/vertx/dq7yTgEECAE

--- a/src/main/java/io/vertx/ext/jdbc/impl/actions/JDBCUpdate.java
+++ b/src/main/java/io/vertx/ext/jdbc/impl/actions/JDBCUpdate.java
@@ -35,7 +35,7 @@ import static io.vertx.ext.jdbc.impl.actions.JDBCStatementHelper.*;
  */
 public class JDBCUpdate extends AbstractJDBCAction<UpdateResult> {
 
-  private static final Pattern regex = Pattern.compile(".*insert.*", Pattern.CASE_INSENSITIVE + Pattern.MULTILINE);
+  private static final Pattern regex = Pattern.compile("(^|\\s)insert(\\s|$)", Pattern.CASE_INSENSITIVE + Pattern.MULTILINE);
 
   private final String sql;
   private final JsonArray in;
@@ -48,7 +48,7 @@ public class JDBCUpdate extends AbstractJDBCAction<UpdateResult> {
 
   @Override
   protected UpdateResult execute(Connection conn) throws SQLException {
-    final boolean returKeys = regex.matcher(sql).matches();
+    final boolean returKeys = regex.matcher(sql).groupCount() == 2;
     try (PreparedStatement statement = conn.prepareStatement(sql, returKeys ? Statement.RETURN_GENERATED_KEYS : Statement.NO_GENERATED_KEYS)) {
       fillStatement(statement, in);
 

--- a/src/main/java/io/vertx/ext/jdbc/impl/actions/JDBCUpdate.java
+++ b/src/main/java/io/vertx/ext/jdbc/impl/actions/JDBCUpdate.java
@@ -26,6 +26,7 @@ import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
+import java.util.regex.Pattern;
 
 import static io.vertx.ext.jdbc.impl.actions.JDBCStatementHelper.*;
 
@@ -33,6 +34,8 @@ import static io.vertx.ext.jdbc.impl.actions.JDBCStatementHelper.*;
  * @author <a href="mailto:nscavell@redhat.com">Nick Scavelli</a>
  */
 public class JDBCUpdate extends AbstractJDBCAction<UpdateResult> {
+
+  private static final Pattern regex = Pattern.compile(".*insert.*", Pattern.CASE_INSENSITIVE + Pattern.MULTILINE);
 
   private final String sql;
   private final JsonArray in;
@@ -45,21 +48,26 @@ public class JDBCUpdate extends AbstractJDBCAction<UpdateResult> {
 
   @Override
   protected UpdateResult execute(Connection conn) throws SQLException {
-    try (PreparedStatement statement = conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
+    final boolean returKeys = regex.matcher(sql).matches();
+    try (PreparedStatement statement = conn.prepareStatement(sql, returKeys ? Statement.RETURN_GENERATED_KEYS : Statement.NO_GENERATED_KEYS)) {
       fillStatement(statement, in);
 
       int updated = statement.executeUpdate();
       JsonArray keys = new JsonArray();
 
       // Create JsonArray of keys
-      try (ResultSet rs = statement.getGeneratedKeys()) {
-        if (rs != null) {
-          while (rs.next()) {
-            Object key = rs.getObject(1);
-            if (key != null) {
-              keys.add(convertSqlValue(key));
+      if (returKeys) {
+        try (ResultSet rs = statement.getGeneratedKeys()) {
+          if (rs != null) {
+            while (rs.next()) {
+              Object key = rs.getObject(1);
+              if (key != null) {
+                keys.add(convertSqlValue(key));
+              }
             }
           }
+        } catch (SQLException e) {
+          // do not crash if no permissions
         }
       }
 

